[TECH][ADR] Nommer les contraintes dans la BDD en cas de conflit (PIX-5048) (#4483)
1024pix/pix: docs/adr/0033-utiliser-noms-contraintes-metiers-lisibilite.md

# 1. Nom des contraintes sur la base PG

Date : 2022-06-01

## État

En cours

## Contexte

Nous sommes amenés à générer des contraintes d'unicité entre plusieurs colonne pour différents contexte métier. PG limite le nom des contraintes à 63 caractères.

> La majorité des contraintes sont générés par `knex` (90% des cas).

Reviewer: ```suggestion
La majorité des contraintes sont générées par `knex` (90% des cas).
```

Actuellement lorsque nous crééons une contrainte spécifique, nous les appelons par le nom des colonnes qui la compose. Sauf que dans certains cas nous atteignons la limite des 63 caractères.

> A savoir que peu importe le nom que nous donnons à notre contraintes elle sera tronqué si elle dépasse cette limite.

Reviewer: ```suggestion
A savoir que peu importe le nom que nous donnons à notre contrainte, elle sera tronqué si elle dépasse cette limite.
```

Nous avons eu deux incidents côté Prescription / Certification. Une contrainte utilisant des colonnes quasiment similaire mais une fois tronqué, les noms de contraintes étaient identique.

### Solution 1 : Nommer les contraintes d'une manière métier et non technique

> Au lieu de nommer les contraintes par le noms des colonnes qui la compose, nous suggérons de les nommer par l'intention métier qu'elle assure.

Reviewer: ```suggestion
Au lieu de nommer les contraintes par le nom des colonnes qui la compose, nous suggérons de les nommer par l'intention métier qu'elle assure.
```

##### Exemple non contractuel

```sql
`CREATE UNIQUE INDEX <name> ON "campaign-participations" ("campaignId", "organizationLearnerId" ) WHERE "isImproved" IS FALSE AND "deletedAt" IS NULL AND "deletedBy" IS NULL;`
```

Une contrainte sur ces colonnes là s'écrirait

- Avec la convention knex (94 caractères)
`'campaign_participations_campaignId_organizationLearnerId_isImproved_deletedAt_deletedBy_unique'`

- En utilisation le contexte métier (35 caractères)
`'one_active_participation_by_learner'`

```js
const OLD_UNIQUE_ORGANIZATION_ID_NATIONAL_APPRENTICE_ID_CONSTRAINT_NAME =
  'schooling_registrations_organizationid_nationalapprenticeid_uni';

const NEW_UNIQUE_ORGANIZATION_ID_NATIONAL_APPRENTICE_ID_CONSTRAINT_NAME =
  'organization_learners_organizationid_nationalapprenticeid_unique';
```

**Avantage(s):**

- Permet d'avoir une lecture simple de ce que la contrainte fait.

> - Si des évolutions métier font que la contrainte doit contenir une nouvelle colonne, pas besoin de renommer la contrainte pour contenir la nouvelle colonne
> - On limite la possibilité de dépassement des 63 caractères
> - plus lisible dans le cas d'erreur en produtction
> - reduction des risques de nommer des contraintes identique
> - ne pas devoir renommer des contraintes parceque la colonne/table a changé de nom

Reviewer: ```suggestion
- Si des évolutions métier font que la contrainte doit contenir une nouvelle colonne, pas besoin de renommer la contrainte pour contenir la nouvelle colonne.
- On limite la possibilité de dépassement des 63 caractères.
- Plus lisible dans le cas d'erreur en production.
- Réduction des risques de nommer des contraintes identique.
- Ne pas devoir renommer des contraintes parce que la colonne/table a changé de nom.
```

**Inconvénient(s):**

> - Gymnastique pas forcément inné pour les développeurs

Reviewer: ```suggestion
- Gymnastique pas forcément innée pour les développeurs
```

- Nouvelle habitude à prendre 
- Pas de renommage des anciennes contraintes

### Solution 2 : Recompiler PG afin d'augmenter le max size des nom de contraintes

La longueur maximale du nom de contrainte est définie dans la constante NAMEDATALEN d'un fichier header C, donc sa valeur peut être modifiée, à condition de recompiler le code source .

https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS

Recompiler PG en augmentant la constante `max size` permettrait de gérer des nom de contraintes plus longs.

**Avantage(s):**

- laisser knex gérer le nommage des contraintes 
- problème de longueur "résolu"

**Inconvénient(s):**

- hébergement PaaS (nous n'avons pas la main sur cette variable d'environnement)
- recompiler PG

### Solution 3 : Générer les clés par hash

Générer le nom de la contrainte avec un hash

**inconvénient(s):** 

- hash non symétrique, donc prend du temps pour identifier la contrainte en cas de problème (se connecter à la BDD)

### Conclusion

> La solution 2 est écarté car PGSql est hébergé par un tiers, nous ne pouvons donc effectuer l'action requise pour augmenter le max size des contraintes

Reviewer: ```suggestion
La solution 2 est écartée car PGSql est hébergé par un tiers, nous ne pouvons donc effectuer l'action requise pour augmenter le max size des contraintes
```

# 1. Nom des contraintes sur la base PG

Date : 2022-06-01

## État

En cours

## Contexte

> Nous sommes amenés à générer des contraintes d'unicité entre plusieurs colonne pour différents contexte métier. PG limite le nom des contraintes à 63 caractères.

Reviewer 2: `colonnes` et `contextes` ^^

La majorité des contraintes sont générées par `knex` (90% des cas).

> Actuellement lorsque nous crééons une contrainte spécifique, nous les appelons par le nom des colonnes qui la compose. Sauf que dans certains cas nous atteignons la limite des 63 caractères.

Reviewer 2: un petit `é` en trop dans `créons` ^^

A savoir que peu importe le nom que nous donnons à notre contrainte elle sera tronqué si elle dépasse cette limite.

Nous avons eu deux incidents côté Prescription / Certification. Une contrainte utilisant des colonnes quasiment similaire mais une fois tronqué, les noms de contraintes étaient identique.

##### Exemple non contractuel

> Deux contraintes d'unicité défini avec ces colonnes :

Reviewer 2: `définie` ou `définies` peut-être, je ne sais pas ? 
Je perds mon français !

`'campaign_participations_campaignId_organizationLearnerId_isImproved_deletedAt_deletedBy_unique'`
`'campaign_participations_campaignId_organizationLearnerId_isImproved_participantExternalId_unique'`

> Les noms qu'elles auront en base pour respecter la limite des 63 caractères sera identique, la seconde migration serait en échec:

Reviewer 2: J'adore pinailler, on peut ne pas changer ici : je propose un espace entre `échec` et `:`

`'campaign_participations_campaignId_organizationLearnerId_isI'`

### Solution 1 : Nommer les contraintes d'une manière métier et non technique

Au lieu de nommer les contraintes par le nom des colonnes qui la compose, nous suggérons de les nommer par l'intention métier qu'elle assure.

##### Exemple non contractuel

```sql
`CREATE UNIQUE INDEX <name> ON "campaign-participations" ("campaignId", "organizationLearnerId" ) WHERE "isImproved" IS FALSE AND "deletedAt" IS NULL AND "deletedBy" IS NULL;`
```

> Une contrainte sur ces colonnes là s'écrirait

Reviewer 2: `colonnes-là`

- Avec la convention knex (94 caractères)
`'campaign_participations_campaignId_organizationLearnerId_isImproved_deletedAt_deletedBy_unique'`

> - En utilisation le contexte métier (35 caractères)

Reviewer 2: `En utilisant` ou `Avec l'utilisation` ?

`'one_active_participation_by_learner'`

```js
const OLD_UNIQUE_ORGANIZATION_ID_NATIONAL_APPRENTICE_ID_CONSTRAINT_NAME =
  'schooling_registrations_organizationid_nationalapprenticeid_uni';

const NEW_UNIQUE_ORGANIZATION_ID_NATIONAL_APPRENTICE_ID_CONSTRAINT_NAME =
  'organization_learners_organizationid_nationalapprenticeid_unique';
```

**Avantage(s):**

- Permet d'avoir une lecture simple de ce que la contrainte fait.

> - Si des évolutions métier font que la contrainte doit contenir une nouvelle colonne, pas besoin de renommer la contrainte pour contenir la nouvelle colonne.

Reviewer 2: `métiers`

- On limite la possibilité de dépassement des 63 caractères.
- Plus lisible dans le cas d'erreur en production.

> - Réduction des risques de nommer des contraintes identique.

Reviewer 2: `identiques`

- Ne pas devoir renommer des contraintes parce que la colonne/table a changé de nom.

**Inconvénient(s):**

- Gymnastique pas forcément innée pour les développeurs
- Nouvelle habitude à prendre 
- Pas de renommage des anciennes contraintes

### Solution 2 : Recompiler PG afin d'augmenter le max size des nom de contraintes

> La longueur maximale du nom de contrainte est définie dans la constante NAMEDATALEN d'un fichier header C, donc sa valeur peut être modifiée, à condition de recompiler le code source .

Reviewer 2: Retrait de l'espace entre `source` et `.`

Pour ma culture G : c'est quoi un header C ?

Author: c'est un fichier du code source de pgsql. qui contient des constants afin de définir les paramètres de base lors de la compilation du logiciel

https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS

Recompiler PG en augmentant la constante `max size` permettrait de gérer des nom de contraintes plus longs.

**Avantage(s):**

- laisser knex gérer le nommage des contraintes 
- problème de longueur "résolu"

**Inconvénient(s):**

- hébergement PaaS (nous n'avons pas la main sur cette variable d'environnement)
- recompiler PG

### Solution 3 : Générer les clés par hash

Générer le nom de la contrainte avec un hash

**inconvénient(s):** 

- hash non symétrique, donc prend du temps pour identifier la contrainte en cas de problème (se connecter à la BDD)

### Conclusion

La solution 2 est écartée car PGSql est hébergé par un tiers, nous ne pouvons donc effectuer l'action requise pour augmenter le max size des contraintes

> La solution 3 est écarté pour une question de lisibilité/identification de la source du problème

Reviewer 2: `écartée`

> La solution 1 semble la plus adapté à notre contexte.

Reviewer 2: `adaptée`

# 1. Nom des contraintes sur la base PG

Date : 2022-06-01

## État

En cours

## Contexte

Nous sommes amenès à générer des contraintes d'unicité entre plusieurs colonne pour différents contexte métier. PG limite le nom des contraintes a 63 caractères.

Actuellement nous appelons nos contraintes par le noms des colonnes qui la compose. Sauf que dans certains cas nous atteignons la limite des 63 caractères. Peu importe le nom que nous donnons à notre contraintes elle sera tronqué si nous dépassons cette limite.

> Nous avons deux incidents côté Prescription / Accès. Une contrainte utilisant portant sur des colonnes quasiment similaire mais une fois tronqué, les nom de contraintes étaient identique.

Author: certif

> ### Solution : Nommer les contraintes d'une manière métier et non technique

Author: ajouter la solution compilation pg pour augmenter le max size

# 1. Nom des contraintes sur la base PG

Date : 2022-06-01

## État

En cours

## Contexte

Nous sommes amenés à générer des contraintes d'unicité entre plusieurs colonne pour différents contexte métier. PG limite le nom des contraintes à 63 caractères.

> Actuellement nous appelons nos contraintes par le nom des colonnes qui la compose. Sauf que dans certains cas nous atteignons la limite des 63 caractères. 

Reviewer 3: Ajouter le fait que le nom des contrainte est généré par `knex` dans 90% des cas

A savoir que peu importe le nom que nous donnons à notre contraintes elle sera tronqué si elle dépasse cette limite.

> Nous avons eu deux incidents côté Prescription / Certification. Une contrainte utilisant des colonnes quasiment similaire mais une fois tronqué, les noms de contraintes étaient identique.

Reviewer 3: Un exemple serait le bienvenu ici

> 

Reviewer 3: Ajouter la solution: générer le nom de la contrainte avec un hash
inconvénient: hash non symétrique, donc prend du temps pour identifier la contrainte en cas de problème (se connecter à la BDD)

### Solution 1 : Nommer les contraintes d'une manière métier et non technique

> Au lieu de nommer les contraintes par le noms des colonnes qui la compose. Nous suggérons de les nommer par l'application métier qu'elle induit.

Reviewer 3: ```suggestion
Au lieu de nommer les contraintes par le noms des colonnes qui la compose, nous suggérons de les nommer par l'intention métier qu'elle assure.
```

### Solution 2 : Recompiler PG afin d'augmenter le max size des nom de contraintes

> 

Reviewer 3: La longueux maximale du nom de contrainte est définie dans la constante `NAMEDATALEN` d'un fichier header C, donc sa valeur peut être modifiée, à condition de recompiler le code source .
https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS

> 

Reviewer 3: ```suggestion
La longueux maximale du nom de contrainte est définie dans la constante `NAMEDATALEN` d'un fichier header C, donc sa valeur peut être modifiée, à condition de recompiler le code source .
https://www.postgresql.org/docs/current/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS
```

> Recompiler PG permettrait d'augmenter le max size afin de ne plus avoir de limite de sizing. (Solution directement écarté, PG Hébergé par scalingo PaaS ce qui ne nous permet pas d'utiliser cette solution)

Reviewer 3: Classer en avantages/inconvénients, voir `docs/adr/0001-enregistrer-les-decisions-concernant-l-architecture.md`
L'avantage ici est de ne pas avoir à nommer les contraintes nous-mêmes, mais de laisser knex les générer

> Recompiler PG permettrait d'augmenter le max size afin de ne plus avoir de limite de sizing. (Solution directement écarté, PG Hébergé par scalingo PaaS ce qui ne nous permet pas d'utiliser cette solution)

Reviewer 3: ```suggestion
Recompiler PG en augmentant la constante `max size` permettrait de gérer des nom de contraintes plus longs.

Cette solution est directement écartée, l'instance PG est fournie par scalingo en BDaaS, ils n'accepteront pas de dournir une version de PostgreSQL recompilée)
```

##### Exemple non contractuel

```js
`CREATE UNIQUE INDEX :name: ON :participationsTable: (:campaignId:, :organizationLearnerId: ) WHERE :isImproved: IS FALSE AND :deletedAt: IS NULL AND :deletedBy: IS NULL;`

const ISIMPROVED_COLUMN = 'isImproved';
const DELETEDAT_COLUMN = 'deletedAt';
const DELETEDBY_COLUMN = 'deletedBy';
const ORGANIZATIONLEARNERID_COLUMN = 'organizationLearnerId';
const CAMPAIGNID_COLUMN = 'campaignId';
```

Une contrainte sur ces colonnes là s'écrirait

```js
//with column naming

> const NEW_CONSTRAINT_NAME = 'campaignId_organizationLearnerId_isImproved_deletedAt_deletedBy_unique'; // 70 characters

Reviewer 3: C'est bizarre de vori du Js parce qu'on a pas l'instruction `knex` qui effectue l'action
Je propose de ne pas utiliser de Js et d'expliquer avec du texte
```suggestion
const NEW_CONSTRAINT_NAME = 'campaignId_organizationLearnerId_isImproved_deletedAt_deletedBy_unique'; // 70 characters
```

> const NEW_CONSTRAINT_NAME = 'campaignId_organizationLearnerId_isImproved_deletedAt_deletedBy_unique'; // 70 characters

Reviewer 3: C'est bizarre de voir du Js parce qu'on a pas l'instruction `knex` qui effectue l'action
Je propose de ne pas utiliser de Js et d'expliquer avec du texte

//with context naming
const NEW_CONSTRAINT_NAME = 'one_active_participation_by_learner'; // 35 characters
```

> On éviterait de devoir renommer des contraintes parceque la colonne / table à changé de nom

Reviewer 3: Si je comprend bien, c'est un avantage.
Si c'est le cas, à remonter dans Solution 1 / avantages

```js
const OLD_UNIQUE_ORGANIZATION_ID_NATIONAL_APPRENTICE_ID_CONSTRAINT_NAME =
  'schooling_registrations_organizationid_nationalapprenticeid_uni';

const NEW_UNIQUE_ORGANIZATION_ID_NATIONAL_APPRENTICE_ID_CONSTRAINT_NAME =
  'organization_learners_organizationid_nationalapprenticeid_unique';
```

### Conclusion

**Avantage(s):**

- Permet d'avoir une lecture simple de ce que la contrainte fait.
- Si des évolutions métier font que la contrainte doit contenir une nouvelle colonne, pas besoin de renommer la contrainte pour contenir la nouvelle colonne
- On limite la possibilité de dépassement des 63 caractères
- plus lisible dans le cas d'erreur en produtction
- reduction des risques de nommer des contraintes identique

**Inconvénient(s):**

- Gymnastique pas forcément inné pour les développeurs
- Nouvelle habitude à prendre 

> - Pas de renommage des anciennes contraintes

Reviewer 3: Le remonter dans la Solution 1, et citer ici ce qui nous décide  prendre celle-ci plutôt que la 2

## Décision

> Suite à la réunion du 01/06/2022, il a été statué que rien n'empêchait d'écrire des contraintes définissant son contexte.

Reviewer 3: ```suggestion
Utiliser la solution n°1, en nommant les contraintes 
- manuellement
- en explicitant l'intention métier

On ne modifie pas le nom des contraintes déjà présentes.

```

## Conséquences

> On peut déjà utiliser ce nommage dans l'application

Reviewer 3: ```suggestion
Nommer manuellement les contraintes (utiliser `knex.raw`)
```
